File: 040_Combination_Sum_II/040_1.java
class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        Deque<Integer> tmp = new LinkedList<>();
        Arrays.sort(candidates);
        dfs(res, tmp, candidates, target, 0);
        return res;
    }

    private void dfs(List<List<Integer>> res, Deque<Integer> tmp, int[] candidates, int target, int start) {
        if (target < 0) return;
        if (target == 0) {
            res.add(new ArrayList<>(tmp));
            return;
        }

        for (int i = start; i < candidates.length; i++) {
            if (target < candidates[i]) break;
            if (i > start && candidates[i] == candidates[i - 1]) continue;
            tmp.add(candidates[i]);
            dfs(res, tmp, candidates, target - candidates[i], i + 1);
            tmp.removeLast();
        }
    }
}

File: 040_Combination_Sum_II/040_1.py
from typing import List, Dict

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        tmp = []
        candidates.sort()

        def dfs(tmp, target, start):
            if target < 0: 
                return
            if target == 0:
                res.append(tmp)
                return
            
            for i in range(start, len(candidates)):
                if target < candidates[i]:
                    break
                if i > start and candidates[i] == candidates[i - 1]:
                    continue
                dfs(tmp + [candidates[i]], target - candidates[i], i + 1)
        
        dfs(tmp, target, 0)
        return res
